lib/utils.js

```diff
@@ -416,20 +416,20 @@ function joinExpressions(expressions) {
   return SPDX.normalize(joinedExpressionString)
 }
 
-function normalizeLicenseExpression(licenseExpression, logger) {
-  if (!licenseExpression) return null
+function normalizeLicenseExpression(rawLicenseExpression, logger) {
+  if (!rawLicenseExpression) return null
 
-  const licenseVisitor = rawLicenseExpression => {
-    const mappedLicenseExpression = scancodeMap.get(rawLicenseExpression)
-    const licenseExpression = mappedLicenseExpression ? mappedLicenseExpression : rawLicenseExpression
-
-    return SPDX.normalizeSingle(licenseExpression)
+  const licenseVisitor = licenseExpression => {
+    return scancodeMap.get(licenseExpression) || SPDX.normalizeSingle(licenseExpression)
   }
 
-  const parsed = SPDX.parse(licenseExpression, licenseVisitor)
+  // parse() checks for LicenseRef- and other special types of expressions before calling the visitor
+  // therefore use the mapped license expression as an argument if it was found
+  const mappedLicenseExpression = scancodeMap.get(rawLicenseExpression)
+  const parsed = SPDX.parse(mappedLicenseExpression || rawLicenseExpression || '', licenseVisitor)
   const result = SPDX.stringify(parsed)
```

Comment:
hm... Look up by scancodeMap.get(rawLicenseExpression) helps to convert a leaf expression node to a mapped value : e.g. `afpl-9.0` can be converted to `LicenseRef-scancode-afpl-9.0`.  If our rawLicenseExpression is `afl-1.1 AND afpl-9.0`, there will still be problems.
```
describe('normalizeLicenseExpression', () => {
  it('should normalize license', () => {
    const expression = 'MIT AND GPL-3.0'
    const result = utils.normalizeLicenseExpression(expression)
    expect(result).to.eq('MIT AND GPL-3.0')
  })
  it('should normalize single licenseRef', () => {
    const expression = 'afpl-9.0'
    const result = utils.normalizeLicenseExpression(expression)
    expect(result).to.eq('LicenseRef-scancode-afpl-9.0')
  })
  it('should normalize license and licenseRef', () => {
    const expression = 'afl-1.1 AND afpl-9.0'
    const result = utils.normalizeLicenseExpression(expression)
    expect(result).to.eq('AFL-1.1 AND LicenseRef-scancode-afpl-9.0')// This one fails with 'AFL-1.1 AND NOASSERTION' 
  })
})
```

Reply:
I looked up `ALF-1.1` and confirmed that it is an SPDX registered license.  `ALPL-9.0` is not, so the expectation is correctly written.  The question is why the second part of the last test was incorrectly assigned `NOASSERTION`.

Reply:
[Third commit](https://github.com/clearlydefined/service/pull/1148/commits/ace7e0526623b0b8af82e71225ce1f44a34aad07) adds tests and processing for complex licenses.  This fixes the issue identified in this thread.  

Thanks @qtomlinson for the tests.  Made it very easy to locate and fix the problem.